datastream-kafka-connector/src/main/java/com/linkedin/datastream/connectors/kafka/AbstractKafkaBasedConnectorTask.java

```diff
@@ -109,6 +109,9 @@
   protected Consumer<?, ?> _consumer;
   protected final Set<TopicPartition> _consumerAssignment = new HashSet<>();
 
+  // TopicPartitions which have seen exceptions on send. Access to this map must be synchronized.
+  private final Map<TopicPartition, Exception> _sendFailureTopicPartitionExceptionMap = new HashMap<>();
```

Comment:
You might have mentioned this to me before, but isnt a ConcurrentHashMap useful here?

Reply:
So there is this one place where I iterate over all the elements and then clear the map. I wanted that block to be synchronized as a whole, which is why I decided to use HashMap instead. That block can be contended with the callbacks on code paths were we skip flush() between calls to poll(). Due to this my worry is that in the middle of iterating, if a new entry gets added due to a send callback completion, what if i miss processing it? Since I clear the map after it, we might miss rewinding some topic partitions.

Documentation for Concurrent collections seem to indicate that there is weak consistency while iterating (https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/package-summary.html#Weakly):

`Most concurrent Collection implementations (including most Queues) also differ from the usual java.util conventions in that their Iterators and Spliterators provide weakly consistent rather than fast-fail traversal:

they may proceed concurrently with other operations
they will never throw ConcurrentModificationException
they are guaranteed to traverse elements as they existed upon construction exactly once, and may (but are not guaranteed to) reflect any modifications subsequent to construction.`